1단계 - 엔티티 매핑 리뷰 요청드립니다!

KwakTaeMin opened the pull request:
안녕하세요 리뷰어님
jpa를 사용하고 테스트 코드를 짜는게 다소 어색하고 부족한 것 같아요ㅠ
jpa테스트에 대한 컨벤션이나 주의할 점 같은 건 없는지 또는 좋은 테스트 방법이 있다면 소개해주시면 감사하겠습니다.
리뷰 잘 부탁드려요!
감사합니다.

KwakTaeMin committed:
docs : 요구사항 정리

KwakTaeMin committed:
feat : hibiernate 수정
- sql 로그 보기 수정
- Mysql Dialect 사용

KwakTaeMin committed:
feat : Answer, DeleteHistory, Question, User
- @Entity , @ID, @Repository 추가
- 기본 TestCode 잘 돌 수 있도록 초기화

Commit:
feat : Answer Entity 매핑

Commit:
feat : DeleteHistory Entity 매핑

Commit:
feat : Question Entity 매핑

Commit:
feat : User Entity 매핑

Commit:
test : Answer Test 코드 작성 중

Commit:
test : AnswerRepository CRUD 테스트

Commit:
feat : Entity에 기본생성자 추가
- Answer, DeleteHistory, Question, User

Commit:
refactor : AnswerRepositoryTest 리펙토링
- AnswerTest에 있던 static final Answer 코드 AnswerRepositoryTest 파일로 이동

Commit:
test : Answer 테스트
- 생성자 테스트
- Question Null validate 테스트
- Wrtier Null validate 테스트

Commit:
test : DeleteHistoryRepositoryTest 테스트
- CRD 테스트

Commit:
style : 모든 클래스 정렬 포멧 수정

Commit:
test : DeleteHistoryTest 생성 테스트

Commit:
style : DeleteHistoryTest @DisplayName 추가

Commit:
refactor : AnswerRepositoryTest의 A1,A2 다시 AnswerTest로 이동
- 통일성을 위해 수정

Commit:
test : QuestionRepositoryTest 테스트
- CRUD 테스트

Commit:
test : Question 테스트
- 객체 생성 테스트

Commit:
test : User 테스트
- 생성 및 벨리데이션 통과 테스트

Commit:
test : UserRepositoryTest 테스트
- CRUD 테스트

Commit:
style : 포멧 수정

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,127 @@
-                ", deleted=" + deleted +
-                '}';
-    }
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	@Column
+	private Long id;
+	@Column
+	private Long writerId;
+	@Column
+	private Long questionId;
+	@Lob
+	@Column
+	private String contents;
+	@Column(nullable = false)
+	private boolean deleted = false;
+	@Temporal(TemporalType.TIMESTAMP)
+	@Column(nullable = false)
+	private Date createAt = new Date();
+	@Temporal(TemporalType.TIMESTAMP)
+	@Column
+	private Date updateAt;
```
가독성이.. 무척 떨어지네요. 개행을 해주세요!

JunilHwang replied:
그리고 `@Column`의 `name` 속성을 통해서 column의 이름을 스네이크 표기법으로 지정해주세요!
ex) `@Column(name = "created_at")`

JunilHwang replied:
그리고 createdAt, updatedAt, id 등은 전체 Entity에서 공통으로 사용 중인데, 이를 상속해서 사용하도록 하면 어떨까요?

KwakTaeMin replied:
개행, 컬럼 네임 지정, 공통된 부분은 상속하여서 사용할 수 있도록 수정할게요!

JunilHwang commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,44 +3,71 @@
+	@Column
+	private Long deletedById;
+	@Column
+	private LocalDateTime createDate;
```
앞선 리뷰와 동일합니다.

JunilHwang commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,83 +1,125 @@
+	@Column(nullable = false)
+	private Date createAt = new Date();
+	@Column
+	private Date updateAt;
```
앞선 리뷰와 동일합니다

JunilHwang commented on src/main/java/qna/domain/User.java:
```diff
@@ -1,122 +1,167 @@
+@Table(
+	name = "user",
+	uniqueConstraints =
+	@UniqueConstraint(columnNames = {"userId"})
```
userId에 직접 지정해도 괜찮을 것 같습니다!

KwakTaeMin replied:
오 네 컬럼에 바로 지정할 수 있는 방법이 있군요! 좋은 피드백 감사합니다!!

JunilHwang reviewed (changes requested):
안녕하세요 태민님! 엔티티 매핑 미션 잘 진행해주셨네요 👍
몇 가지 리뷰 남겼으니 확인 후에 다시 피드백 요청 부탁드립니다 🙇‍♂️

> jpa를 사용하고 테스트 코드를 짜는게 다소 어색하고 부족한 것 같아요ㅠ jpa테스트에 대한 컨벤션이나 주의할 점 같은 건 없는지 또는 좋은 테스트 방법이 있다면 소개해주시면 감사하겠습니다.

음.. 테스트 컨벤션에 대해서는 딱히 들어본적이 없네요 ㅎㅎ
테스테에 대한 관점은 사실 JPA 테스트나 일반 단위테스트나 크게 다를게 없다고 생각합니다.
JPA 자체가 Database를 객체지향을 잘 표현하기 위해서 사용하는 것이기 때문에 일반 단위 테스트처럼 작성하면 된다고 생각합니다.

대신 신경써야 할 부분이 바로 데이터 연동인데, `QnAServiceTest` 처럼 아예 목 데이터를 사용하는 경우도 있고, 혹은 `@DataJpaTest` 처럼 DB에 Connect하여 테스트 하는 방법도 있습니다. 목킹으로 할 경우에 속도는 빠르지만 손이 굉장히 많이 가는 편이고, 직접 DB에 연결할 경우에 환경설정만 하면 되지만 실제 DB를 사용하지 않도록 해야하며 테스트 DB로 H2를 사용하는 경우가 많기 때문에 MySQL과 미세하게 다른 부분이 있을 수 있습니다. 이런 것들은 사실 직접 경험해봐야 와닿는 것 같아요!

이번 JPA 과정에서 직접 경험해보시길 권합니다 ㅎㅎ

KwakTaeMin committed:
feat : BaseEntity 구현
- 공통된 id, createAt, updateAt과 같은 속성을 모아 중복 코드 제거를 위한 구현

KwakTaeMin committed:
refactor : Answer, Question, User, DeleteHistory 리펙토링
- Answer, Question, User BaseEntity 상속하여 중복 코드 제거
- Entity 개행 및 column 이름 스네이크 표기법으로 작성
- 가독성을 위해 개행

KwakTaeMin committed:
fix : AnswerRepositoryTest 테스트 코드 버그 수정

KwakTaeMin committed:
style : Optimize style

KwakTaeMin commented:
질문에 아주 상세한 답변해주셔서 정말 감사합니다.
앞으로도 모르는게 있을 때 많이 고민하고 많이 질문하도록 하겠습니다.
잘 부탁드립니다!!

KwakTaeMin committed:
style : BaseEntity 개행

JunilHwang reviewed (approved):
개행이 하나도 안되어있네요 ㅠㅠ 일단 Step1은 여기서 마무리하겠습니다! 바로 Step2 진행해주세요~

KwakTaeMin commented:
옴...개행이 
@Column
private Long id;

@Column
private String name;

요런 느낌이 아닌가요?ㅜㅜ
